### ISSUE TITLE
Python: Fix UUID representation

### ISSUE THREAD

**Issue by Fokko** [apache/iceberg#8248]
### Feature Request / Improvement

Currently, we represent a UUID literal as `UUIDLiteral[Literal[UUID]]`, but we want to change this to `UUIDLiteral[Literal[bytes]]`.

Probably currently also tables that are partitioned by UUIDs will fail, because the comparison will be between an UUID and bytes. 

What we can do, is create a table using PyIceberg:

```python
from pyiceberg.catalog import load_catalog
from pyiceberg.schema import Schema
from pyiceberg.types import UUIDType, NestedField

schema = Schema(
    NestedField(
        field_id=1, name="c1", field_type=UUIDType(), required=False
    ),
)

cat.create_table(
    identifier="default.uuid",
    schema=schema
)
```

And then add some data:

```sql
insert into uuid VALUES('102cb62f-e6f8-4eb0-9973-d9b012ff0967');
```

Would also be great to have tests around partitioned tables, but it looks like that's not possible today: https://github.com/apache/iceberg/issues/8247

cc @JonasJ-ap @HonahX I know that you were looking into this as well.

### Query engine

None

**Comment by HonahX** [comment]
More context:

if we applied a row filter on UUID col, 
```python
unpartitioned_uuid = catalog.load_table("default.test_uuid_and_fixed_unpartitioned")
arrow_table_eq = unpartitioned_uuid.scan(row_filter="uuid_col == '102cb62f-e6f8-4eb0-9973-d9b012ff0967'").to_arrow()
```
it fails with
```
pyarrow.lib.ArrowInvalid: Could not convert UUID('102cb62f-e6f8-4eb0-9973-d9b012ff0967') with type UUID: did not recognize Python value type when inferring an Arrow data type
```
raised here:
https://github.com/apache/iceberg/blob/f5f543a54ff7460648bb864f4f06a29eb28938b9/python/pyiceberg/io/pyarrow.py#L451-L454
https://github.com/apache/iceberg/blob/f5f543a54ff7460648bb864f4f06a29eb28938b9/python/pyiceberg/io/pyarrow.py#L480-L481
because UUID is a `fixed_binary[16]` in pyarrow but the UUID literal stores UUID in its value

**Comment by Fokko** [comment]
@HonahX We need to store the `bytes` instead of the `UUID`, are you interested in fixing this?

**Comment by HonahX** [comment]
@Fokko Yes! I just uploaded a draft PR https://github.com/apache/iceberg/pull/8267 for further discussion. This PR verifies that changing the literal value can solve the issue. I will try to see if I can add more tests to it.

Since pyiceberg use the released spark-iceberg-runtime to create the test env, we may not be able to add tables partitioned by uuid until the next release.

**Comment by raphaelauv** [comment]
@Fokko @HonahX thanks for the feature but I have a problem using it 

I can't insert data in an iceberg table that have a column with UUID type
```python
from pyiceberg.schema import Schema
from pyiceberg.types import NestedField, UUIDType
import polars as pl

import uuid


id = uuid.uuid4()

SCHEMA = Schema(
    NestedField(1, "id", UUIDType(), required=True),
)

df = pl.DataFrame({}).with_columns([pl.lit(id.bytes).alias("id")])

df = df.to_arrow()
df = df.cast(target_schema=SCHEMA.as_arrow())
```

![image](https://github.com/apache/iceberg/assets/10202690/7b75f2fa-7715-44ea-8b97-637d905aa422)

do you have any idea ? thanks

**Comment by Fokko** [comment]
@raphaelauv This looks like a new issue, can you open up a new ticket? Thanks for including some example code there 👍

**Comment by raphaelauv** [comment]
Thanks @Fokko ! Should I create the issue in the repo py-iceberg ?
